File: api/sdks/python/tradewave/_http.py
"""HTTP transport for the TradeWave SDK.

A thin wrapper over a ``requests.Session`` that:
  - attaches the ``Authorization: Bearer <api_key>`` header and a User-Agent,
  - parses the JSON error envelope into the typed exception hierarchy,
  - retries automatically on 429 and 5xx with exponential backoff + jitter,
    honoring ``Retry-After`` when present,
  - exposes any ``X-RateLimit-*`` headers from the last response.

Only ``requests`` is used - no other third-party dependency.
"""

from __future__ import annotations

import random
import time
from typing import Any, Dict, Mapping, Optional

import requests

from . import __version__
from .errors import (
    ConnectionError as TWConnectionError,
    RateLimitError,
    TimeoutError as TWTimeoutError,
    TradeWaveError,
    error_for_status,
)

DEFAULT_BASE_URL = "https://api.tradewave.ai/v1"
DEFAULT_TIMEOUT = 30.0
DEFAULT_MAX_RETRIES = 3
_BACKOFF_BASE = 0.5  # seconds; doubled each attempt
_BACKOFF_CAP = 30.0  # never sleep longer than this between retries
_USER_AGENT = f"tradewave-python/{__version__}"


def _parse_retry_after(value: Optional[str]) -> Optional[float]:
    """Parse a Retry-After header (seconds form) into a float, else None."""
    if not value:
        return None
    try:
        return max(0.0, float(value))
    except (TypeError, ValueError):
        # HTTP-date form is allowed by the spec but the API uses seconds; if we
        # can't parse it, fall back to the computed backoff.
        return None


class HTTPClient:
    """Owns the ``requests.Session`` and the request/retry loop."""

    def __init__(
        self,
        api_key: str,
        *,
        base_url: str = DEFAULT_BASE_URL,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
        session: Optional[requests.Session] = None,
        user_agent: Optional[str] = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max(0, int(max_retries))
        self._owns_session = session is None
        self.session = session or requests.Session()
        self.session.headers.update(
            {
                "Authorization": f"Bearer {api_key}",
                "User-Agent": user_agent or _USER_AGENT,
                "Accept": "application/json",
            }
        )
        # Rate-limit headers from the most recent response (X-RateLimit-*).
        self.last_rate_limit: Dict[str, str] = {}

    # -- lifecycle ---------------------------------------------------------

    def close(self) -> None:
        if self._owns_session:
            self.session.close()

    def __enter__(self) -> "HTTPClient":
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()

    # -- core request loop -------------------------------------------------

    def request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Mapping[str, Any]] = None,
        json: Optional[Any] = None,
    ) -> Any:
        """Send a request and return the parsed JSON body.

        Retries on 429 and 5xx (and on transient network errors) up to
        ``max_retries``; raises a :class:`TradeWaveError` subclass on a
        terminal non-2xx response. Non-retryable 4xx (other than 429) raise
        immediately.
        """
        url = f"{self.base_url}/{path.lstrip('/')}"
        clean_params = (
            {k: v for k, v in params.items() if v is not None} if params else None
        )

        attempt = 0
        while True:
            try:
                resp = self.session.request(
                    method,
                    url,
                    params=clean_params,
                    json=json,
                    timeout=self.timeout,
                )
            except requests.Timeout as exc:
                if attempt < self.max_retries:
                    self._sleep_backoff(attempt, None)
                    attempt += 1
                    continue
                raise TWTimeoutError(f"Request to {url} timed out") from exc
            except requests.ConnectionError as exc:
                if attempt < self.max_retries:
                    self._sleep_backoff(attempt, None)
                    attempt += 1
                    continue
                raise TWConnectionError(
                    f"Failed to connect to {url}: {exc}"
                ) from exc
            except requests.RequestException as exc:
                raise TradeWaveError(f"Request to {url} failed: {exc}") from exc

            self._capture_rate_limit(resp.headers)

            if 200 <= resp.status_code < 300:
                return self._parse_json(resp)

            # Decide whether to retry.
            retry_after = _parse_retry_after(resp.headers.get("Retry-After"))
            retryable = resp.status_code == 429 or 500 <= resp.status_code < 600
            if retryable and attempt < self.max_retries:
                self._sleep_backoff(attempt, retry_after)
                attempt += 1
                continue

            # Terminal: build and raise the typed error.
            raise self._error_from_response(resp, retry_after)

    # -- helpers -----------------------------------------------------------

    def _sleep_backoff(self, attempt: int, retry_after: Optional[float]) -> None:
        """Sleep before the next retry: honor Retry-After, else exp backoff."""
        if retry_after is not None:
            delay = retry_after
        else:
            delay = min(_BACKOFF_CAP, _BACKOFF_BASE * (2 ** attempt))
        # Full jitter on top of the computed delay to avoid thundering herds.
        delay += random.uniform(0, _BACKOFF_BASE)
        time.sleep(delay)

    def _capture_rate_limit(self, headers: Mapping[str, str]) -> None:
        captured = {
            k: v for k, v in headers.items() if k.lower().startswith("x-ratelimit")
        }
        if captured:
            self.last_rate_limit = captured

    @staticmethod
    def _parse_json(resp: requests.Response) -> Any:
        if not resp.content:
            return {}
        try:
            return resp.json()
        except ValueError as exc:
            raise TradeWaveError(
                f"Expected JSON but got non-JSON response (HTTP {resp.status_code})",
                status=resp.status_code,
            ) from exc

    def _error_from_response(
        self, resp: requests.Response, retry_after: Optional[float]
    ) -> TradeWaveError:
        body: Optional[Dict[str, Any]] = None
        code: Optional[str] = None
        message = f"HTTP {resp.status_code}"
        try:
            parsed = resp.json()
            if isinstance(parsed, dict):
                body = parsed
                envelope = parsed.get("error")
                if isinstance(envelope, dict):
                    code = envelope.get("code")
                    message = envelope.get("message") or message
                elif isinstance(parsed.get("message"), str):
                    message = parsed["message"]
        except ValueError:
            text = (resp.text or "").strip()
            if text:
                message = text[:500]

        return error_for_status(
            resp.status_code,
            message=message,
            code=code,
            response=body,
            retry_after=retry_after,
        )

File: api/sdks/typescript/src/http.ts
/**
 * Zero-dependency HTTP layer for the TradeWave SDK.
 *
 * Uses the global `fetch` (Node 18+ / 22, Deno, browsers). Adds:
 *  - Bearer auth + a small User-Agent
 *  - configurable timeout via AbortController
 *  - automatic exponential backoff with jitter on 429 and 5xx, honoring Retry-After
 *  - the typed error hierarchy from errors.ts
 *
 * The ML graceful nudge (HTTP 200 with requires/reason or ml:null) is plain success
 * data and is returned as-is, never treated as an error.
 */

import { errorForStatus, RateLimitError, TradeWaveError } from './errors.js';

export const SDK_VERSION = '0.1.0';
export const DEFAULT_BASE_URL = 'https://api.tradewave.ai/v1';
const USER_AGENT = `tradewave-typescript/${SDK_VERSION}`;

export interface HttpClientOptions {
  apiKey: string;
  baseUrl?: string;
  /** Per-request timeout in milliseconds. Default 30000. */
  timeoutMs?: number;
  /** Max automatic retries on 429 / 5xx. Default 3. */
  maxRetries?: number;
  /** Optional custom fetch (for testing). Defaults to the global fetch. */
  fetch?: typeof fetch;
}

export interface RequestOptions {
  /** Query parameters. undefined / null values are dropped. */
  query?: Record<string, string | number | boolean | undefined | null>;
  /** JSON request body (for POST). */
  body?: unknown;
  method?: 'GET' | 'POST';
}

interface RateLimitInfo {
  limit?: number;
  remaining?: number;
  reset?: number;
}

/** A parsed response: the JSON payload plus rate-limit metadata read off the headers. */
export interface ApiResponse<T> {
  data: T;
  status: number;
  rateLimit: RateLimitInfo;
  /** ml_remaining_today header / body hint surfaced for convenience when present. */
  headers: Headers;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Parse a Retry-After header (seconds or HTTP-date) into milliseconds, or undefined. */
function parseRetryAfter(value: string | null): number | undefined {
  if (!value) return undefined;
  const seconds = Number(value);
  if (Number.isFinite(seconds)) return Math.max(0, seconds * 1000);
  const date = Date.parse(value);
  if (!Number.isNaN(date)) return Math.max(0, date - Date.now());
  return undefined;
}

function readRateLimit(headers: Headers): RateLimitInfo {
  const num = (name: string): number | undefined => {
    const raw = headers.get(name);
    if (raw == null) return undefined;
    const n = Number(raw);
    return Number.isFinite(n) ? n : undefined;
  };
  return {
    limit: num('X-RateLimit-Limit'),
    remaining: num('X-RateLimit-Remaining'),
    reset: num('X-RateLimit-Reset'),
  };
}

export class HttpClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeoutMs: number;
  private readonly maxRetries: number;
  private readonly fetchImpl: typeof fetch;

  constructor(options: HttpClientOptions) {
    if (!options.apiKey) {
      throw new TradeWaveError(
        'A TradeWave API key is required. Pass apiKey or set TRADEWAVE_API_KEY.',
      );
    }
    this.apiKey = options.apiKey;
    this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, '');
    this.timeoutMs = options.timeoutMs ?? 30000;
    this.maxRetries = options.maxRetries ?? 3;
    const f = options.fetch ?? globalThis.fetch;
    if (typeof f !== 'function') {
      throw new TradeWaveError(
        'No global fetch available. Use Node 18+ / 22, or pass a fetch implementation.',
      );
    }
    // Bind so it is not called with the wrong receiver.
    this.fetchImpl = f.bind(globalThis);
  }

  private buildUrl(path: string, query?: RequestOptions['query']): string {
    const url = new URL(this.baseUrl + (path.startsWith('/') ? path : `/${path}`));
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value === undefined || value === null) continue;
        url.searchParams.set(key, String(value));
      }
    }
    return url.toString();
  }

  /** Exponential backoff with full jitter, in milliseconds. */
  private backoffMs(attempt: number): number {
    const base = Math.min(1000 * 2 ** attempt, 20000);
    return Math.random() * base;
  }

  async request<T>(path: string, options: RequestOptions = {}): Promise<ApiResponse<T>> {
    const method = options.method ?? 'GET';
    const url = this.buildUrl(path, options.query);

    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
      Accept: 'application/json',
      'User-Agent': USER_AGENT,
    };
    let bodyText: string | undefined;
    if (options.body !== undefined) {
      headers['Content-Type'] = 'application/json';
      bodyText = JSON.stringify(options.body);
    }

    let lastError: unknown;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.timeoutMs);
      let response: Response;
      try {
        response = await this.fetchImpl(url, {
          method,
          headers,
          body: bodyText,
          signal: controller.signal,
        });
      } catch (err) {
        clearTimeout(timer);
        // Network error / timeout: retry like a transient failure.
        lastError = err;
        const aborted = err instanceof Error && err.name === 'AbortError';
        const msg = aborted
          ? `Request to ${path} timed out after ${this.timeoutMs}ms`
          : `Network error calling ${path}: ${(err as Error).message}`;
        if (attempt < this.maxRetries) {
          await sleep(this.backoffMs(attempt));
          continue;
        }
        throw new TradeWaveError(msg, { cause: err });
      } finally {
        clearTimeout(timer);
      }

      const rateLimit = readRateLimit(response.headers);

      if (response.ok) {
        const data = (await this.parseJson(response)) as T;
        return { data, status: response.status, rateLimit, headers: response.headers };
      }

      // Error path. Read the envelope { error: { code, message } } when present.
      const envelope = await this.parseJson(response).catch(() => undefined);
      const errObj =
        envelope && typeof envelope === 'object' && 'error' in envelope
          ? (envelope as { error?: { code?: string; message?: string } }).error
          : undefined;
      const code = errObj?.code;
      const message =
        errObj?.message ?? `TradeWave API error ${response.status} on ${path}`;
      const retryAfter = parseRetryAfter(response.headers.get('Retry-After'));

      const retryable = response.status === 429 || response.status >= 500;
      if (retryable && attempt < this.maxRetries) {
        // Honor Retry-After when present, else backoff with jitter.
        const waitMs =
          retryAfter !== undefined ? retryAfter : this.backoffMs(attempt);
        await sleep(waitMs);
        lastError = errorForStatus(response.status, message, {
          code,
          retryAfter: retryAfter !== undefined ? retryAfter / 1000 : undefined,
        });
        continue;
      }

      throw errorForStatus(response.status, message, {
        code,
        retryAfter: retryAfter !== undefined ? retryAfter / 1000 : undefined,
      });
    }

    // Exhausted retries on a retryable status.
    if (lastError instanceof RateLimitError || lastError instanceof TradeWaveError) {
      throw lastError;
    }
    throw new TradeWaveError(`Request to ${path} failed after retries`, {
      cause: lastError,
    });
  }

  private async parseJson(response: Response): Promise<unknown> {
    const text = await response.text();
    if (!text) return undefined;
    try {
      return JSON.parse(text);
    } catch (err) {
      throw new TradeWaveError(
        `Failed to parse JSON response (status ${response.status}).`,
        { status: response.status, cause: err },
      );
    }
  }
}
